Client handling refactoring

vsh51 opened the pull request.

vsh51 committed:
Lock-free approach adjusted for thread-safe queue

vsh51 committed:
refactor(core): redesign structure to follow OOP principles

- Refactored ClientHandling structure to improve adherence to OOP
- Updated key components and logic for better modularity and readability. Mainly CommandHandler changed
- Work in progress, not just little logic is finalized

vsh51 committed:
feat(ClientSession): Continued refactoring structure

Moved logic up to handling data

vsh51 committed:
feat(server): remove CommandHandler

- Deleted CommandHandler component files
- Adjusted CMakeLists.txt files

vsh51 committed:
refactor(client-session): update ClientSession and SmtpRequest logic

- Finished moving CommandHandler into ClientSession
- Updated ClientSession.cpp and SmtpRequest.cpp with new logic and optimizations
- Refactored newly created class
- Slightly changed Data parsing logic

This refactor enhances the maintainability of SMTP server

hiddensudo commented on include/Server/include/ClientSession/src/SmtpRequest.cpp:
```diff
@@ -0,0 +1,163 @@
+
+std::string RequestParser::ExtractBody(std::string data)
+{
+    std::size_t start_pos = data.find("\r\n\r\n") + 4;
```
\r\n can be put into a constant, for example, static constexpr const char* CRLF = “\r\n”;

The value '4' is a magic number. It would be better to define a constant for clarity.

hiddensudo commented on include/Server/include/ClientSession/src/SmtpRequest.cpp:
```diff
@@ -0,0 +1,163 @@
+    }
+    else if (request.find("NOOP") == 0)
+    {
+        smtp_request.command = SmtpCommand::NOOP;
+    }
+    else if (request.find("STARTTLS") == 0)
+    {
+        smtp_request.command = SmtpCommand::STARTTLS;
+    }
+    else if (request.find("REGISTER") == 0)
+    {
+        smtp_request.command = SmtpCommand::REGISTER;
+    }
+    else if (request.find("AUTH") == 0)
+    {
+        smtp_request.command = SmtpCommand::AUTH;
+    }
+    else if (request.find("MAIL FROM") == 0)
+    {
+        smtp_request.command = SmtpCommand::MAILFROM;
+    }
+    else if (request.find("RCPT TO") == 0)
+    {
+        smtp_request.command = SmtpCommand::RCPTTO;
+    }
+    else if (request.find("DATA") == 0)
+    {
+        smtp_request.command = SmtpCommand::DATA;
+    }
+    else if (request.find("RSET") == 0)
+    {
+        smtp_request.command = SmtpCommand::RSET;
+    }
+    else if (request.find("QUIT") == 0)
+    {
+        smtp_request.command = SmtpCommand::QUIT;
+    }
+    else {
+        throw std::runtime_error("Invalid SMTP command.");
+    }
```
It might be better to use a std::unordered_map to improve the readability and maintainability of the command parse logic.

You can also consider other options besides std::unordered_map.

hiddensudo commented on include/Server/include/ClientSession/src/SmtpRequest.cpp:
```diff
@@ -0,0 +1,163 @@
+
+std::string RequestParser::ExtractSubject(std::string data)
+{
+    std::size_t start_pos = data.find("Subject: ") + 9;
+    std::size_t end_pos = data.find("\r\n", start_pos);
```
\r\n can be put into a constant.

The value '9' is a magic number.

hiddensudo commented on include/Server/include/ClientSession/src/SmtpRequest.cpp:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    // Remove the last period from the body
+    std::size_t last_pos = data.find("\r\n.", start_pos);
```
\r\n can be put into a constant

hiddensudo commented on include/Server/include/ClientSession/include/ClientSession.h:
```diff
@@ -0,0 +1,98 @@
+#ifndef ClientState_H
+#define ClientState_H
```
According to the code convention, it might be better to call this header CLIENT_STATE_H.

hiddensudo reviewed (commented):
The code looks good, but pay attention to the comments.

zikkend commented on include/Server/include/ClientSession/src/ClientSession.cpp:
```diff
@@ -0,0 +1,768 @@
+        HadleStartTls(request);
+        m_current_state = ClientState::STARTTLS_SENT;
+    }
+    else
```
After EHLO command client does not necessarily send STARTTLS. Client can even do not send STARTTLS at all

zikkend commented on include/Server/include/ClientSession/src/ClientSession.cpp:
```diff
@@ -0,0 +1,768 @@
+    {
+        HandleRegister(request);
+        m_current_state = ClientState::AUTH_SENT;
+    }    else
```
Client can send EHLO command after STARTTLS

zikkend commented on include/Server/include/ClientSession/src/ClientSession.cpp:
```diff
@@ -0,0 +1,768 @@
+    Logger::LogDebug("Entering ClientSession::HandleRcptToSentState");
+    Logger::LogTrace("ClientSession::HandleRcptToSentState parameter: const SmtpRequest reference: " + request.data);
+
+    if (request.command == SmtpCommand::DATA)
```
maybe it is better to use switch in all ` HandleState`

zikkend commented on include/Server/include/ClientSession/src/ClientSession.cpp:
```diff
@@ -0,0 +1,768 @@
+
+/******************/
+/* State handlers */
+/******************/
```
Almost all states should handle NOOP command, but i don't see this

vsh51 committed:
style(headers): standardize header guards and format comments

- Updated header guards to uppercase format.
- Improved formatting and consistency of comments.
